### PR TITLE
Microsoft.DigitalTwins - add stable data-plane API version 2022-05-31

### LINKED ISSUE #18792: [DigitalTwins] REST API Review
**Service name:** Azure Digital Twins

**Key contact for the review:** Sebastian Herzig-Patel (@sjiherzig)

**Whether this is a new or existing API:** new API version, but no changes to existing APIs

**Whether or not your service already been deployed:** no

**A brief description of the material to be reviewed/changed:**
This API version derives from 2021-06-30-preview and the swagger has no changes other than omitting the enumeration literals for the api-version parameter. There are no behavioral changes w.r.t. 2021-06-30-preview. This API version is mainly introduced in order to provider a stable API version for the GA release of the ADT Data History feature and the accompanying support for setting the "sourceTime" property.

### PR DESCRIPTION
**Pull request by sjiherzig**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [x] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by sjiherzig** [comment]
To help drive the review, here is a summary of the changes from the previous version (2021-06-30-preview):
* Updated API version header
* Removed enumeration literals from api-version parameter, and converted to string per past guidance
* Added "sourceTime" usage to examples

**Comment by sjiherzig** [comment]
Note: I don't know what to do about the ModelValidation failure. We have used an "object" return type in the past, since the API response greatly depends on a model defined by the customer.

**Comment by weidongxu-microsoft** [comment]
@jianyexi for ModelValidation failure. At least the error message does not make sense for INVALID_TYPE. `Expected type {0} but found type {1}.` but it does not say anything for expected type.

**Comment by jianyexi** [comment]
> @jianyexi for ModelValidation failure. At least the error message does not make sense for INVALID_TYPE. `Expected type {0} but found type {1}.` but it does not say anything for expected type.

@raych1  to confirm if it's a bug

File: specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json [hunk 1]
```diff
@@ -0,0 +1,1622 @@
...
+      "required": true,
+      "x-ms-parameter-location": "method"
+    },
+    "api-version": {
+      "name": "api-version",
+      "in": "query",
+      "description": "The requested API version.",
+      "required": true,
+      "type": "string",
+      "minLength": 10
```
Comment (weidongxu-microsoft) [inline specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json:1555]: You might just use minLength=1
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/types.json#L636

**Review by weidongxu-microsoft — APPROVED** [review]
There is no change since 2020-10-31 except removing enum from api-verison
```
    "api-version": {
      "name": "api-version",
      "in": "query",
      "description": "The requested API version.",
      "required": true,
      "type": "string",
      "enum": [
        "2020-10-31"
      ]
    },
```

File: specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json [hunk 2]
```diff
@@ -0,0 +1,1622 @@
...
+            "headers": {
+              "ETag": {
+                "description": "Weak Etag.",
+                "type": "string"
+              }
+            }
+          },
+          "202": {
+            "description": "Asynchronous Success"
+          },
```
Comment (weidongxu-microsoft) [inline specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json:418]: these appear to be long-running operation

File: specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json [hunk 3]
```diff
@@ -0,0 +1,1622 @@
...
+        "code": {
+          "description": "Service specific error code which serves as the substatus for the HTTP error code.",
+          "type": "string",
+          "readOnly": true
+        },
+        "message": {
+          "description": "A human-readable representation of the error.",
+          "type": "string",
+          "readOnly": true
+        },
```
Comment (weidongxu-microsoft) [inline specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json:1463]: nit, error and code and message should be required.

**Review by weidongxu-microsoft — APPROVED** [review]

File: specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json (hunk 2 above)
Comment (sjiherzig) [reply specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json:418]: We are currently not returning a 202 with this API - I believe a previous review of an upcoming preview API version recommended removing this response code from the swagger. Would that be considered a breaking change? (We don't actually ever return it).

Comment (sjiherzig) [reply specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json:418]: Actually, per a previous review (https://github.com/Azure/azure-rest-api-specs-pr/issues/5257), I think it's OK to remove.

Comment (sjiherzig) [reply specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json]: I've removed them. This was discussed in a previous API review (of an upcoming preview API).

File: specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json (hunk 3 above)
Comment (sjiherzig) [reply specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json]: Done

File: specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json (hunk 1 above)
Comment (sjiherzig) [reply specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json]: Done

**Comment by weidongxu-microsoft** [comment]
"required" in schema should be done like this: https://github.com/Azure/azure-rest-api-specs/blob/6d6b7c15deab7ae76ed51ca9c2f20c13cca57065/specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json#L1245-L1248

**Comment by sjiherzig** [comment]
> "required" in schema should be done like this:
> 
> https://github.com/Azure/azure-rest-api-specs/blob/6d6b7c15deab7ae76ed51ca9c2f20c13cca57065/specification/digitaltwins/data-plane/Microsoft.DigitalTwins/stable/2022-05-31/digitaltwins.json#L1245-L1248

This has now been addressed.

**Comment by sjiherzig** [comment]
One thing to note - I've removed the "readyOnly" attribute from the error response model parameters because one of the automated validation tools complained that a required property should not be readyonly. This doesn't quite make sense to me - doesn't readonly set only the context whether it is from a response or a request? Should / can this be fixed?

**Comment by weidongxu-microsoft** [comment]
@sjiherzig
My fault. OpenAPI 2.0 does not allow specifying both "readonly" and "required" to property. (see https://swagger.io/specification/v2/ on "readonly")

Our guide https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#handling-errors and some SDK weight more on "required", as in this particular case, "readonly" of the whole "Error" schema can be deduced (as "Error" is not used as input in any operation).

In this case, maybe not adding the "required" is the base choice, as it keep it same as your last GA api-version. But do keep it on "ErrorResponse".

**Comment by weidongxu-microsoft** [comment]
@sjiherzig 

Please also note that the guideline asks for the "x-ms-error-code" header on error response.
https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#handling-errors

Please do evaluate whether it can be included in this api-version (it requires backend change). If yes, please support it; if not, please open a backlog item in your service to add it in next api-version.

**Review by weidongxu-microsoft — APPROVED** [review]
LGTM

skip ModelValidation CI, as it is intended for digitaltwin having response schema of type=object

**Comment by sjiherzig** [comment]
Hi @weidongxu-microsoft - this is now ready to be merged!

**Comment by weidongxu-microsoft** [comment]
@sjiherzig

Do we have an issue relates to the PR?

I didn't remember since it is a long time ago. It is possible you do not have the issue and meeting notes, as there is no API change since last version.

**Comment by sjiherzig** [comment]
@weidongxu-microsoft Yes, there is a GitHub issue for this PR: https://github.com/Azure/azure-rest-api-specs/issues/18792 .

There are no outstanding action items for the PR. The review was done asynchronously.

**Pull request merged**